Initial Qt6 port

JamiKettunen opened the pull request:
Builds as tested on Qt 6.8.0. Current status: [`asteroid-calculator` with patches](https://github.com/JamiKettunen/asteroid-calculator/tree/qt6) runs!
<img src="https://github.com/user-attachments/assets/2cf72012-e130-444a-9161-ad0c25337645" width="360" />

My AsteroidOS package integration work for [Chimera Linux](https://chimera-linux.org/) which this PR is a part of along with all other patches can be found at https://github.com/JamiKettunen/cports/commits/hybris/; I'll likely update this description when I have more stuff of interest to share..

JamiKettunen committed:
Fix building against Qt6

JamiKettunen committed:
Adapt all QML to use JS functions with formal parameters

Fixes runtime warnings such as:
  qt.qml.context: qrc:/org/asteroid/controls/qml/BorderGestureArea.qml:160:5 Parameter "mouse" is not declared. Injection of parameters into signal handlers is deprecated. Use JavaScript functions with formal parameters instead.

As an alternative the following also would work:
  function onPressed(mouse) { ... }

JamiKettunen committed:
[WIP] Generate asteroidcontrolsplugin qrc via cmake

Re-use the already existing listing of controls used for documentation
generation as well.

WIP because this will be replaced by qt_add_qml_module()

JamiKettunen committed:
Port {Circular,}Spinner fragmentShader to Qt6 (qsb)

JamiKettunen committed:
org.controls.asteroid -> org.asteroid.controls

This is a typo in all the existing docs.

FlorentRevest commented:
Oh nice! This is a very welcome step 😄 do you have broader plans regarding Qt6 adoption in Asteroid or was that a one-off experiment ? And if so, is there any way we could help or support you with that ?

JamiKettunen commented:
I do want to get all of the AsteroidOS UI and apps running on top of Qt6 with Halium/libhybris yes, but we'll see how large this ends up being as yet another project I've decided to undertake for some reason :rofl:

JamiKettunen committed:
Port AsteroidApp to Qt6

JamiKettunen committed:
generate-desktop.sh: Stop using unportable GNU grep -P

This doesn't work on Chimera Linux which uses the FreeBSD grep(1); adds
a new build-time dependency on xmllint which I'm not sure how to feel
about yet.

JamiKettunen commented:
So I got `AsteroidApp` and such going now on https://github.com/JamiKettunen/cports/commits/hybris/, testing [`asteroid-calculator`](https://github.com/AsteroidOS/asteroid-calculator) launched from Konsole under KDE Plasma 6 desktop I can get it to launch *if* I replace the https://github.com/AsteroidOS/asteroid-calculator/blob/master/src/CMakeLists.txt#L1 call to `add_library()` with `add_executable()` and then run the resulting `/usr/lib/asteroid-calculator` executable:
<img src="https://github.com/user-attachments/assets/2cf72012-e130-444a-9161-ad0c25337645" width="360" />
If I don't do that it just `SEGFAULT`s with these same weird addresses in the backtrace with no further info (launching through the `/usr/bin/asteroid-calculator` wrapper which calls the mapplauncher booster through invoker gives even less detailed errors and still crashes in seemingly the same way?)
```
$ lldb -- /usr/lib/asteroid-calculator.so
(lldb) target create "/usr/lib/asteroid-calculator.so"
Current executable set to '/usr/lib/asteroid-calculator.so' (x86_64).
(lldb) r
Process 20576 launched: '/usr/lib/asteroid-calculator.so' (x86_64)
Process 20576 stopped
* thread #1, name = 'asteroid-calcul', stop reason = signal SIGSEGV: invalid permissions for mapped object (fault address: 0x7ffff7ff9000)
    frame #0: 0x00007ffff7ff9000 asteroid-calculator.so
->  0x7ffff7ff9000: jg     0x7ffff7ff9047
```
Attempting to run `/usr/bin/asteroid-calculator` (`invoker --single-instance --type=qt6 /usr/lib/asteroid-calculator.so`) outputs:
```
invoker: error: method call failed: org.freedesktop.DBus.Error.ServiceUnknown: The name org.sailfishos.sailjaild1 was not provided by any .service files
invoker: warning: application (pid=8884) exit(1) signal(0)
```
Full log from `/usr/libexec/mapplauncherd/booster-generic -v` running while attempting to launch `/usr/bin/asteroid-calculator`: https://paste.c-net.org/kz9r1zj3rn7s

Either the mapplauncher stuff is just broken or it's perhaps some weird (lib)gcc/stdlibc++ thing the functionality depends on maybe? I'm using full llvm/libc++/compiler-rt/libunwind toolchain

JamiKettunen committed:
Make Mapplauncherd support optional

It's worth revisiting if this makes sense to keep as a dependency with
some repeatable benchmarks where a quantifiable difference may or may
not be observed when running this on a watch.

If this is to be removed cmake/FindMapplauncherd_qt*.cmake additionally
should also be gone.

JamiKettunen committed:
cmake: Let's not pretend 3.6.0 is a minimum supported version

src/controls/CMakeLists.txt already makes use of the following:
https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#list-transformations

JamiKettunen committed:
Adapt to modern Qt6 QML import path conventions

Let's adopt QTP0001 already and move all per-application resources under
qrc:/qt/qml/AsteroidApp/

JamiKettunen committed:
Warn on missing translations

To fix the following observed build warning:
src/app/asteroidapp.cpp:80:13: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
   80 |             translator->load(QLocale(), appName(), ".", "/usr/share/translations", ".qm");
      |             ^~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

JamiKettunen committed:
Start adding QML type annotations for AOT cachegen

JamiKettunen committed:
asteroidapp: qmake: Call generate-desktop.sh with sh instead of bash

It's a POSIX sh script as indicated by its shebang.

JamiKettunen committed:
Generate everything org.asteroid.utils related via CMake

JamiKettunen committed:
[RESIVIT] Use CMake qt_qml_add_module() for org.asteroid.controls

This one (for now at least) retains the manually written plugin instead
of generating one since it also changes the default font instead of just
calling qmlRegisterType for each C++ QML component:

  QGuiApplication::setFont(QFont("Noto Sans"));

If another way to accomplish the same task globally for anything using
org.asteroid.controls is found the same as one commit above for
org.asteroid.utils can also be done here and controls_plugin.* can be
done away with.
